torch/optim/swa_utils.py

```diff
@@ -0,0 +1,262 @@
+import torch
+import math
+from torch.nn import Module
+from copy import deepcopy
+from torch.optim.lr_scheduler import _LRScheduler
+
+
+class AveragedModel(Module):
+    r"""Implements averaged model for Stochastic Weight Averaging (SWA).
+
+    Stochastic Weight Averaging was proposed in `Averaging Weights Leads to
+    Wider Optima and Better Generalization`_ by Pavel Izmailov, Dmitrii
+    Podoprikhin, Timur Garipov, Dmitry Vetrov and Andrew Gordon Wilson
+    (UAI 2018).
+
+    AveragedModel class creates a copy of the provided module :attr:`model`
+    on the device :attr:`device` and allows to compute running averages of the 
+    parameters of the :attr:`model`.
+
+    Arguments:
+        model (torch.nn.Module): model to use with SWA
+        device (torch.device, optional): if provided, the averaged model will be
+            stored on the :attr:`device` 
+        avg_fn (function, optional): the averaging function used to update 
+            parameters; the function must take in the current value of the 
+            :class:`AveragedModel` parameter, the current value of :attr:`model`
+            parameter and the number of models laready averaged; if None, 
+            equally weighted average is used (default: None)
+
+    Example:
+        >>> loader, optimizer, model, loss_fn = ...
+        >>> swa_model = torch.optim.swa_utils.AveragedModel(model)
+        >>> # swa_model computes an equally-weighted average of the weights
+        >>> # You can use custom averaging functions with `avg_fun` parameter
+        >>> ema_avg = lambda p_avg, p, n_avg: 0.1 * p_avg + 0.9 * p
+        >>> ema_model = torch.optim.swa_utils.AveragedModel(model, 
+        >>>                                     avg_fun=ema_avg)
+        >>> # ema_model computes exponential moving averages of the weights
+        >>> scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, 
+        >>>                                     T_max=300)
+        >>> swa_start = 160
+        >>> swa_scheduler = SWALR(optimizer, swa_lr=0.05)
+        >>> for i in range(300):
+        >>>      for input, target in loader:
+        >>>          optimizer.zero_grad()
+        >>>          loss_fn(model(input), target).backward()
+        >>>          optimizer.step()
+        >>>      if i > swa_start:
+        >>>          swa_model.update_parameters(model)
+        >>>          ema_model.update_parameters(model)
+        >>>          swa_scheduler.step()
+        >>>      else:
+        >>>          scheduler.step()
+        >>>
+        >>> # Update bn statistics for the swa_model at the end
+        >>> torch.optim.swa_utils.update_bn(loader, swa_model) 
```

Comment:
Are both `swa_model` and `ema_model` needed in the same example? From our previous discussion, I would have expected:
```python
loader, optimizer, model = ...

# You can use custom averaging functions with `avg_fun` parameter
# If no averaging function is provided, the default is to compute equally-weighted average of the weights
ema_avg = lambda p_avg, p, n_avg: 0.1 * p_avg + 0.9 * p
ema_model = torch.optim.swa_utils.AveragedModel(model, avg_function=ema_avg)

scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=300)
swa_scheduler = SWALR(optimizer, swa_lr=0.05)
swa_start = 160

for i in range(300):
     for input, target in loader:
         optimizer.zero_grad()
         loss_fn(model(input), target).backward()
         optimizer.step()

     if i > swa_start:
         swa_scheduler.step()
         ema_model.update_parameters(model)
     else:
         scheduler.step()

# Update bn statistics for the swa_model at the end
torch.optim.swa_utils.update_bn(loader, ema_model)
```

Reply:
I guess, my preference is to use `swa_model` in the main example: that's the averaging we use in the paper and what we recommend using. Could we have the main example with `swa_model` and then another small example like this?
```python3
# You can use custom averaging functions with `avg_fun` parameter
# If no averaging function is provided, the default is to compute equally-weighted average of the weights
ema_avg = lambda p_avg, p, n_avg: 0.1 * p_avg + 0.9 * p
ema_model = torch.optim.swa_utils.AveragedModel(model, avg_function=ema_avg)
```

Reply:
I have pushed a version where I split the example into two as I suggested above. Please let me know what you think!

Reply:
Ah, I see. The name change between `swa_model` and `ema_model` is a little confusing. Let's use `swa_model` for both.